Data/ByteString/Internal.hs

```diff
@@ -142,11 +174,18 @@ import GHC.Ptr                  (Ptr(..), castPtr)
 -- "Data.ByteString.Char8" it can be interpreted as containing 8-bit
 -- characters.
 --
-data ByteString = PS {-# UNPACK #-} !(ForeignPtr Word8) -- payload
-                     {-# UNPACK #-} !Int                -- offset
+data ByteString = BS {-# UNPACK #-} !(ForeignPtr Word8) -- payload
+                     -- {-# UNPACK #-} !Int                -- offset
                      {-# UNPACK #-} !Int                -- length
     deriving (Typeable)
 
+
+#if __GLASGOW_HASKELL__ >= 800
+pattern PS :: ForeignPtr Word8 -> Int -> Int -> ByteString
+pattern PS fp zero len <- BS fp (((,) 0) -> (zero, len)) where
```

Comment:
Should this be bidirectional, if you want to preserve API for `Data.ByteString.Internal` users?

Reply:
Every consumer that I know of works correctly with the bidirectional version here. 

If it only works one direction then one half of the users of the internal API would break as there are lots of consumers that use it to read bytestrings, but also ones that use it to produce them (like the mmap library, etc.). 

This isn't a perfect match for the API of `Internal`, but a 99.5% kind of thing. What I don't want is for users to be hoist on the horns of the dilemma of either tying themselves to a new version of `bytestring`, which entails dropping support for older GHCs, or littering sometimes 100s of use-sites with `#if MIN_VERSION_bytestring`.

This way if you are a user with a support window that spans over the last 4 or so GHC versions you probably won't notice. . This covers 95% of the consumers of the library. If you only ever use this pattern synonym in one direction, you definitely can't notice at all.

If you have a long enough support window that the pattern synonym doesn't cover the whole range, you should switch to using explicit calls to `to/fromForeignPtr` to avoid being tied to the current version of `bytestring` or having to litter your code with checks for the BS and PS constructors.

The only real way that this goes wrong is for an imaginary subset of users that first compares the pointer, for instance, then goes passes control away to something else that does normal bytestring operations and then finally compares result each `bytestring` `ForeignPtr` against the original to see whether or not they have a substring of what they were given or if they had to produce a new string and then uses that to do something new. 

Even this non-existent use-case can be fixed, BTW, by having them hack their code to compare the `ForeignPtrContents` side of the `ForeignPtr` for equality, which is safer anyways, but they'd have to actually change code, so I'd consider them the 0.5%.